### PR TITLE
Change Basic settings field display; better admin UX

### PR DESCRIPTION
**Pull request by joshcanhelp**
- Refactored Basic settings with new render functions, new descriptions
- Fixed PHP < 5.4 array style for advanced render functions
- Persist tab showing on settings page 
- Universal show/hide JS for switch fields
- Constant for JWKS cache transient name

### CODE REVIEW

File: assets/js/admin.js [hunk 1]
```diff
@@ -37,24 +38,60 @@ jQuery(document).ready(function($) {
...
 
+        $.post(wpa0.ajax_url, postData, function() {
+            $deleteCacheButton.prop( 'disabled', false ).val( wpa0.clear_cache_done );
+        }, 'json');
+    } );
+
+    function localStorageAvailable() {
+        try {
+            var x = '__storage_test__';
+            window.localStorage.setItem(x, x);
```
Comment (lbalmaceda) [inline assets/js/admin.js]: is this try/catch the only way to check for availability? And, can the `removeItem` fail after the first one succeeded?

File: lib/admin/WP_Auth0_Admin_Basic.php [hunk 2]
```diff
@@ -32,153 +32,120 @@ public function init() {
...
-	?>
-      <input type="text" autocomplete="off" name="<?php echo $this->options->get_options_name(); ?>[client_secret]" id="wpa0_client_secret"  <?php if ( !empty( $v ) ) {?>placeholder="Not visible"<?php } ?> />
-      <div class="subelement">
-        <span class="description"><?php echo __( 'Application secret, copy from your application\'s settings in the', 'wp-auth0' ); ?> <a href="https://manage.auth0.com/#/applications" target="_blank">Auth0 dashboard</a>.</span>
-      </div>
-  <?php
+	public function render_client_secret_b64_encoded( $args = array() ) {
+		$this->render_switch( $args['label_for'], $args['opt_name'] );
+		$this->render_field_description(
+			__( 'Enable this if your Client Secret is base64 enabled. ', 'wp-auth0' ) .
```
Comment (lbalmaceda) [inline lib/admin/WP_Auth0_Admin_Basic.php]: base64 _encoded_

File: lib/admin/WP_Auth0_Admin_Basic.php [hunk 3]
```diff
@@ -32,153 +32,120 @@ public function init() {
...
+		printf(
+			' <input type="button" id="auth0_delete_cache_transient" value="%s" class="button button-secondary">',
+			__( 'Delete Cache', 'wp-auth0' )
+		);
+		$this->render_field_description( __( 'JWKS cache expiration in minutes (use 0 for no caching)', 'wp-auth0' ) );
+		if ( $domain = $this->options->get( 'domain' ) ) {
+			$this->render_field_description(
+				sprintf( '<a href="https://%s/.well-known/jwks.json" target="_blank">%s</a>',
+					$domain,
+					__( 'View your JWKS here', 'wp-auth0' )
```
Comment (lbalmaceda) [inline lib/admin/WP_Auth0_Admin_Basic.php]: Would this section show the JWKs contents or what?

File: assets/js/admin.js [hunk 4]
```diff
@@ -37,24 +38,60 @@ jQuery(document).ready(function($) {
...
-        // Hide/Show login method depending on auto login
-        var $loginMethodField = $("#wpa0_auto_login_method").closest("tr");
-        var $autoLoginCheckbox = $("#wpa0_auto_login");
-        if (!$autoLoginCheckbox.prop("checked")) {
-            $loginMethodField.hide();
-        }
+    // Show/hide field for specific switches
+    $('[data-expand!=""]').each( function() {
+        var $thisSwitch = $( this );
+        var $showFieldRow = $( '#' + $thisSwitch.attr( 'data-expand' ) ).closest( 'tr' );
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply assets/js/admin.js]: Length is always there for a jQuery selector. Next line checks that

File: assets/js/admin.js (hunk 1 above)
Comment (joshcanhelp) [reply assets/js/admin.js]: No, but it's the best way:

https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API#Feature-detecting_localStorage

As for `removeItem`, that would be a weird case, not implementing the webStorage interface. But localStorage persist forever and we don't want to store something we're never going to use.

File: lib/admin/WP_Auth0_Admin_Basic.php (hunk 3 above)
Comment (joshcanhelp) [reply lib/admin/WP_Auth0_Admin_Basic.php]: Links to it, yes.

File: lib/admin/WP_Auth0_Admin_Basic.php (hunk 2 above)
Comment (joshcanhelp) [reply lib/admin/WP_Auth0_Admin_Basic.php]: Why did I say "enabled" here? 😆 

Fixed!

File: assets/js/admin.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (lbalmaceda) [reply assets/js/admin.js]: I meant the "data-expand" and then "closest TR".

[the remark this answers is not in the document]
Comment (joshcanhelp) [reply assets/js/admin.js]: Having `data-expand` is what gets you into the `each` here. No `data-expand`, no `each`

File: assets/js/admin.js (hunk 1 above)
Comment (lbalmaceda) [reply assets/js/admin.js]: is this stored under some domain key or can it collide with existing "x" stored keys? In that case, you might want to use some key like "a0_local_storage_assertion"

Comment (joshcanhelp) [reply assets/js/admin.js]: It's per-domain:

https://stackoverflow.com/q/4201239/728480

But you have a good point about that. WP sites could load a million other JS libraries.

**Pull request merged**